### ISSUE TITLE
postfix completion on a closure could probably work better. 

### ISSUE THREAD

**Issue by jhgg** [rust-analyzer/rust-analyzer#11922]
consider the following code:

```rust
let x = || {};
```

This defines x as a closure that returns nothing. However, when doing:

```rust
let x = || {}.box;
```

and selecting the box postfix completion, currently, rust-analyzer does the following:

```rust
let x = || Box::new({});
```

However, what we *probably* want to do in this case is actually:

```rust
let x = Box::new(|| {});
```

Wonder if we should special case this.

**Comment by Veykril** [comment]
Wouldn't it be enough to check if our receiver is a block expression and our parent is a closure expression?

**Comment by jhgg** [comment]
> Wouldn't it be enough to check if our receiver is a block expression and our parent is a closure expression?

Yeah! That's what I'm trying to say, but when we are having the syntax: `|| {}.box`, the parent of the BLOCK_EXPR is a FIELD_EXPR, not a CLOSURE_EXPR.

**Comment by Veykril** [comment]
Ye but I mean isn't it the default for the completion handler to find a `FieldExpr` there, even outside of closures? What I meant was to check the parent of the `FieldExpr` for whether it is a closure. Or do you mean handling the closure as the receiver won't work because that will include the `.whatever` in its text? I see that being annoying but I would imagine one could get around that problem somewhat, though I don't exactly remember how the postfix completions get constructed.

**Comment by jhgg** [comment]
The issue at hand as I understand it is that under the current postfix implementation it tries to capture the child of the FIELD_EXPR as the receiver. In order to make this suggested change work though, we need to capture the parent (CLOSURE_EXPR) and but not the FIELD_EXPR, as that would cause the receiver to be invalid. So we essentially need to find the field expr, remove it, but keep its children. See the AST I posted above for what I mean.

**Comment by Veykril** [comment]
Ye I get that part. What I meant is to special case the behaviour inside of `completions/postfix.rs`. That is we check if the receiver is a `BlockExpr` and then if the second parent of that `BlockExpr`, first parent being the field access expression, is a `ClosureExpr`, do some extra shenanigans so that we apply the things on the `|..| { ... }` part instead, which shouldn't be too difficult to do given the entire code there basically works off of text ranges already instead of nodes.
